mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -1580,6 +1580,48 @@ OpFoldResult ReshapeOp::fold(FoldAdaptor adaptor) {
           llvm::dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
           getResult().getType()))
     return reshapedSource;
+
+  auto source = getSource();
+  auto sourceTy = dyn_cast<RankedTensorType>(source.getType());
+  auto resultTy = dyn_cast<RankedTensorType>(getType());
+
+  if (!sourceTy || !resultTy || sourceTy != resultTy)
+    return {};
+
+  if (auto fromElements = getShape().getDefiningOp<tensor::FromElementsOp>()) {
+    auto elements = fromElements.getElements();
+    bool dynamicNoop =
```

Comment:
Maybe take this condition out and check before? If this is false to start with, then nothing to do?

Reply:
And this should be `resultTy`?

Reply:
It can be but the main idea is that the rank of the type doesn't change. I.e. that the number of elements is same as the input/output rank. I can change to resultTy, but at this point we have already verified that they are the same.

Reply:
Also updated / simplified the loops contents.